### DISCUSSION TITLE
PowerShell is powerful but also slow when it starts, how to improve?

### DISCUSSION THREAD

**Discussion by SlimeNull** [PowerShell/PowerShell#17834] [Q&A]
We know that pwsh is always slow when first starts. It takes about 3sec on my PC. (No profile script, and it takes about 500ms when the second starts)

What did pwsh do? Why it's so slow when first starts? How can we improve its starting speed?

> Known:
> 1. I didn't use the profile script so that might be not the problem
> 2. I have already added the environment variable "POWERSHELL_UPDATECHECK" and set it to "Off"
> 3. I didn't install any modules to pwsh. there is only one module named "PSReadLine"  that is integrated with pwsh.
> 4. I cannot run the scripts that use ngen.exe because there seems no ngen.exe on my PC (At least, I cannot find it)

**Comment by doctordns** [comment]
Regarding point 4, ngen,exe is part of the SDK. 

Have you loaded the .NET SDK?

**Reply by fMichaleczek** [reply]
ngen has been replaced by [ReadyToRun](https://docs.microsoft.com/en-us/dotnet/core/deploying/ready-to-run) for PowerShell 7.2, but it seems disabled for 7.3-Preview (https://github.com/PowerShell/PowerShell/blob/master/PowerShell.Common.props#L140 )

you could try  to disable Telemetry but i'm not sure if there is any benefit in the startup sequence (https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_telemetry?view=powershell-7.2)

**Comment by kasini3000** [comment]
Add a way to lazily autoload command on $profile
- add a command
- does not increment the command, but lazy loads everything in $profile.

https://github.com/PowerShell/PowerShell/issues/17283

"powershell 7 start slow " Many users complain about this, but users don't want to think about loading too much in $profile. 
I hope psteam will take this issue seriously, provide a lazy loading scheme, 
at the best able to output the load times of loaded content for each item in profile.
